File: generated_code/python/solution_1917_2.py
import heapq
from typing import List

class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        # Helper function to calculate the gain in pass ratio if one student is added
        def get_gain(p: int, t: int) -> float:
            return (p + 1) / (t + 1) - p / t

        # Max-heap to always pick the class that benefits the most from an extra student.
        # Python's heapq is a min-heap by default, so we store negative gain to simulate a max-heap.
        max_heap = []
        for p, t in classes:
            gain = get_gain(p, t)
            max_heap.append((-gain, p, t))
        
        heapq.heapify(max_heap)

        # Greedily allocate each extra student to the class with the maximum ratio gain
        for _ in range(extraStudents):
            neg_gain, p, t = heapq.heappop(max_heap)
            p += 1
            t += 1
            new_gain = get_gain(p, t)
            heapq.heappush(max_heap, (-new_gain, p, t))

        # Calculate the final total pass ratio across all classes
        total_ratio = 0.0
        for _, p, t in max_heap:
            total_ratio += p / t

        return total_ratio / len(classes)

File: generated_code/cpp/solution_1917_2.cpp
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        // Lambda to calculate the marginal gain in pass ratio if one student is added
        auto get_gain = [](double pass, double total) {
            return (pass + 1) / (total + 1) - pass / total;
        };

        // Max-heap stores elements as {gain, pass, total}
        // We want the class that gives the maximum gain at the top
        priority_queue<tuple<double, int, int>> max_heap;

        double total_ratio = 0.0;

        // Initialize the heap with the initial gain of each class
        for (const auto& c : classes) {
            int pass = c[0];
            int total = c[1];
            total_ratio += (double)pass / total;
            max_heap.push({get_gain(pass, total), pass, total});
        }

        // Greedily distribute each extra student to the class with the highest potential gain
        for (int i = 0; i < extraStudents; ++i) {
            auto [current_gain, pass, total] = max_heap.top();
            max_heap.pop();

            // Subtract the old ratio from our tracking sum and add the new one
            total_ratio += current_gain;

            // Increment the student counts for this class
            pass++;
            total++;

            // Push the updated class back into the heap with its new potential gain
            max_heap.push({get_gain(pass, total), pass, total});
        }

        // Return the average pass ratio across all classes
        return total_ratio / classes.size();
    }
};